include/sycl/algorithm/rotate_copy.hpp

```diff
@@ -0,0 +1,80 @@
+/* Copyright (c) 2015-2018 The Khronos Group Inc.
+
+   Permission is hereby granted, free of charge, to any person obtaining a
+   copy of this software and/or associated documentation files (the
+   "Materials"), to deal in the Materials without restriction, including
+   without limitation the rights to use, copy, modify, merge, publish,
+   distribute, sublicense, and/or sell copies of the Materials, and to
+   permit persons to whom the Materials are furnished to do so, subject to
+   the following conditions:
+
+   The above copyright notice and this permission notice shall be included
+   in all copies or substantial portions of the Materials.
+
+   MODIFICATIONS TO THIS FILE MAY MEAN IT NO LONGER ACCURATELY REFLECTS
+   KHRONOS STANDARDS. THE UNMODIFIED, NORMATIVE VERSIONS OF KHRONOS
+   SPECIFICATIONS AND HEADER INFORMATION ARE LOCATED AT
+    https://www.khronos.org/registry/
+
+  THE MATERIALS ARE PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+  EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+  MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
+  IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
+  CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
+  TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
+  MATERIALS OR THE USE OR OTHER DEALINGS IN THE MATERIALS.
+
+*/
+
+#ifndef __SYCL_IMPL_ALGORITHM_ROTATE_COPY__
+#define __SYCL_IMPL_ALGORITHM_ROTATE_COPY__
+
+#include <algorithm>
+#include <vector>
+
+namespace sycl {
+namespace impl {
+
+/* rotate_copy.
+ * Implementation of the command group that submits a rotate_copy kernel.
+ */
+template <class ExecutionPolicy, class ForwardIt1, class ForwardIt2>
+ForwardIt2 rotate_copy(ExecutionPolicy &sep, ForwardIt1 first,
+                       ForwardIt1 middle, ForwardIt1 last,
+                       ForwardIt2 result) {
+
+  if (first == last) return result;
+
+  using namespace cl::sycl;
+  using value_type = typename std::iterator_traits<ForwardIt1>::value_type;
+  std::vector<value_type> tmp(first, last);
+  std::copy(first, last, tmp.begin());
+{
+  buffer<value_type> bufI(tmp.cbegin(), tmp.cend()), bufO(tmp.size());
+  bufI.set_final_data(nullptr);
+  bufO.set_final_data(tmp.data());
```

Comment:
Destination of set_final_data can be an output iterator, so you could avoid the extra copy.
From SYCL 1.2.1:
```cpp
The finalData points to where the outcome
of all the buffer processing is going to be
copied to at destruction time, if the buffer
was involved with a write accessor.
Destination can be either an output iterator
or a weak_ptr_class<T>
```

Reply:
I like that suggestion. The thing is, I get a compilation error (no matching member function) with ComputeCpp 1.0.4 if I provide an iterator to set_final_data. For example, the code below:

```
#include <SYCL/sycl.hpp>
  
int main(int argc, char *argv[])
{
  std::vector<int>      vec(128);
  cl::sycl::buffer<int> buf(128);
  buf.set_final_data(vec.begin());
  return 0;
}
```

Reply:
That is a good point, we actually missed this variant on the interface. I've reported this internally and I'll keep you posted. In the meantime, I am happy for this to be merged, and possibly and issue be created to re-add the set_final_data with the Output Iterator once it is there.

Reply:
Sounds good. Let me know about set_final_data and I'll add the change then. Cheers.

Reply:
I had a look in ComputeCpp version 1.1.0, but it seems the issue with set_final_data remains. Are you able to merge?

Reply:
Yes, sure. We are waiting for some more changes into the data model before implementing the missing variant, so we'll re-visit this once its added.